File: php/MinionCollective.php
<?php
/**
* MinionCollective class
* Allows you to interact with the distributed task pool
*/
class MinionCollective{

	/******************
	 * Creates the MinionCollective Object
	 * Takes in timeout, dbName, taskPoolName, mongoURI, and mongooptions
	 * timeout is the maximum execution time before regarded as failure in seconds
	 * dbName is name of the database that will be used
	 * taskPoolName is the name of the collection that will be used
	 * mongoURI is connection string used to connect to MongoDB
	 * mongoOptions is an array containing the options for connecting to MongoDB
	******************/
	function __construct($timeout = 60, $dbName = 'MinionCollective', $taskPoolName = 'queue', $mongoURI = 'mongodb://localhost:27017', $mongoOptions = array()){
		$this->timeout = $timeout;
		$this->taskPoolName = $taskPoolName;
		$this->dbName = $dbName;
		$this->mongoURI = $mongoURI;
		$this->mongoOptions = $mongoOptions;
	}

	/******************
	 * Adds a job into the task pool
	 * Takes in an action, identifier, as well as data concerning the job
	 * Data can be anything (string, array), as long as size of $job is less than 4MB
	 * Returns an array with the results of the insert
	******************/
	public function addJob($action, $id, $data = ''){
		$job = array('action' => $action, 'id' => $id, 'data' => $data, 'status' => 'W', 'last_update' => time());
        $collection = $this->connectCollection();
        return $collection->insert($job);
	}

	/******************
	 * Pulls out the oldest undone job from the task pool
	 * Returns an array representing the job that was inserted
	 * Array has the indexes 'action', 'id', 'data', specified when adding job
	 * In addition, there is a '_id' index to be used when reporting the end of job
	******************/
	public function getJob(){
		$job = array('status' => 'W');
        $modify = array('$set' => array('status' => 'I', 'last_update' => time()));
        $options = array('sort' => array("last_update" => 1));
		$collection = $this->connectCollection();
		return $collection->findAndModify($job, $modify, null, $options);
	}

	/******************
	 * Identical to getJob, except it pulls out jobs that are suspected of having crashed
	******************/	
	public function getExpiredJob($timeout = -1){
		if($timeout == -1){
			$timeout = $this->timeout;
		}
		$job = array('status' => 'I', 'last_update' => array('$lt' => time() - $timeout));
        $modify = array('$set' => array('status' => 'I', 'last_update' => time()));
        $options = array('sort' => array("last_update" => 1));
		$collection = $this->connectCollection();
		return $collection->findAndModify($job, $modify, null, $options);
	}

	/******************
	 * Specifies a job as finished
	 * Returns an array representing the job that was inserted
	******************/
	public function finishJob($id, $new = FALSE){
        $job = array('_id' => $id);
        $modify = array('$set' => array('status' => 'C', 'last_update' => time()));
        $options = array('new' => $new);
        $collection = $this->connectCollection();
        return $collection->findAndModify($job, $modify, NULL, $options);
    }

    /******************
	 * Removes all completed jobs
	******************/
	public function removeJobs(){
        $job = array('status' => 'C');
        $collection = $this->connectCollection();
        $collection->remove($job);
    }

    public function getJobCount(){
    	return $this->connectCollection()->count();
    }

    private function connectDB(){
		return new MongoClient($this->mongoURI, $this->mongoOptions);
	}

	private function connectCollection(){
		$conn = $this->connectDB();
		$collectionName = $this->taskPoolName;
		if(is_a($conn, 'MongoClient')){
			$dbName = $this->dbName;
			return $conn->$dbName->$collectionName;
		}
		else{
			return $conn->$collectionName;
		}
		
		return $this->connectDB()->MinionCollective->queue;
	}

	private $timeout;
	private $taskPoolName;
	private $mongoURI;
	private $mongoOptions;
	private $dbName;
}


?>

File: python/MinionCollective.py
import pymongo;
import mongodb_proxy;
from time import time


class MinionCollective:
	def __init__(self, timeout = 60, dbName = 'MinionCollective', taskPoolName = 'queue', mongoURI = 'mongodb://localhost:27017', replicaSet = None):
		self._timeout = timeout
		self._dbName = dbName
		self._taskPoolName = taskPoolName
		self._mongoURI = mongoURI
		self._replicaSet = replicaSet
		if replicaSet == None:
			self._conn = self._connect_server()
		else:
			self._conn = self._connectReplicaSet()

	def add_job(self, action, id, data = ''):
		job = {
			'action'		: action,
			'id'			: id,
			'data'			: data,
			'status'		: 'W',
			'last_update'	: int(time())
		}
		return self._connect_collection().insert(job)

	def get_job(self, action = None):
		if action == None:
			job = {'status' : 'W'}
		else:
			job = {
				'status' : 'W',
				'action' : action
			}
		modify = {
			'$set' : {
				'status' : 'I',
				'last_update' : int(time())
			}
		}
		return self._connect_collection().find_and_modify(job, modify, False, [('last_update', 1)])

	def get_expired_job(self):
		job = {
			'status' : 'I',
			'last_update' : {
				'$lt' : int(time()) - self._timeout
			}
		}
		modify = {
			'$set' : {
				'status' : 'I',
				'last_update' : int(time())
			}
		}
		return self._connect_collection().find_and_modify(job, modify, False, [('last_update', 1)])

	def finish_job(self, id, newVar = False):
		job = {
			'_id' : id
		}
		modify = {
			'$set' : {
				'status' : 'C',
				'last_update' : int(time())
			}
		}
		return self._connect_collection().find_and_modify(job, modify, False, None, False, new=newVar)

	def remove_jobs(self):
		job = {
			'status' : 'C'
		}
		return self._connect_collection().remove(job)

	def get_job_count(self):
		return self._connect_collection().count()


	def _connect_server(self):
		return mongodb_proxy.MongoProxy(pymongo.MongoClient(self._mongoURI));

	def _connect_replica_set(self):
		return mongodb_proxy.MongoProxy(pymongo.MongoReplicaSetClient(self._mongoURI, self._replicaSet));

	def _connect_collection(self):
		if not self._conn.alive():
			if self._replicaSet == None:
				self._conn = self._connect_server()
			else:
				self._conn = self._connectReplicaSet()
		return self._conn[self._dbName][self._taskPoolName]


